File: leetcode/p00007/p00007.py
class Solution:
    def reverse_recursion(self, x: int) -> int:
        sig = 1
        if x < 0:
            sig = -1
            x *= -1
        
        res = 0
        while x != 0:
            num = x % 10
            res = res*10 + num
            x //= 10
        
        if res > pow(2, 31)-1:
            return 0
        
        return res * sig

    def reverse_string(self, x: int) -> int:
        s = str(x)
        head = 0
        
        r = ""
        if s[head] == '-':
            head += 1
            r += '-'

        for i in range(len(s)-1, head-1, -1):
            r += s[i]
        
        if int(r) > pow(2, 31)-1:
            return 0
        
        return int(r)

if __name__ == '__main__':
    s = Solution()

    assert s.reverse_recursion(123) == 321
    assert s.reverse_recursion(-123) == -321
    assert s.reverse_recursion(120) == 21
    assert s.reverse_recursion(1234567899) == 0

    assert s.reverse_string(123) == 321
    assert s.reverse_string(-123) == -321
    assert s.reverse_string(120) == 21
    assert s.reverse_string(1234567899) == 0

File: leetcode/p00007/p00007.go
package p00007

import (
	"math"
	"strconv"
)

// Time Complexity: O(n)
// Space Complexity: O(n)
func reverse_string(x int) int {
	s := strconv.Itoa(x)

	head := 0
	if s[head] == '-' {
		head++
	}

	r := ""
	for i := len(s) - 1; i >= head; i-- {
		r += string(s[i])
	}

	num, _ := strconv.Atoi(r)
	if num > math.MaxInt32-1 {
		return 0
	}

	if x > 0 {
		return num
	}
	return num * -1
}

// Time Complexity: O(n)
// Space Complexity: O(1)
func reverse_recursion(x int) int {
	sig := 1
	if x < 0 {
		x *= -1
		sig = -1
	}

	res := 0
	for x != 0 {
		num := x % 10
		res = res*10 + num
		x /= 10
	}

	if res > math.MaxInt32-1 {
		return 0
	}

	return res * sig
}
